src/compiler/commandLineParser.ts

```diff
@@ -664,6 +667,7 @@ const commandOptionsWithoutBuild: CommandLineOption[] = [
         affectsEmit: true,
         affectsBuildInfo: true,
         affectsModuleResolution: true,
+        affectsSemanticDiagnostics: true,

```

Comment:
If there is affectsModuleResolution we dont need `affectsSemanticDiagnostics` i think. Is there error which doesnt get refreshed. Because changing jsx option changed module resolution it should update reference graph if any change at all and update the diagnostics.

Reply:
Curiously, the jsx baseline changes for this; is that change incorrect?

Reply:
Specifically, see `updates-emit-on-jsx-option-change.js`.

Reply:
I think the baseline change was not expected in this. If you add a external module to that test case, then it will update the diagnostics for that module.  jsx import is resolved only if:

```ts
if ((getIsolatedModules(options) || isExternalModuleFile)
            && !file.isDeclarationFile) {
            if (options.importHelpers) {
                // synthesize 'import "tslib"' declaration
                imports = [createSyntheticImport(externalHelpersModuleNameText, file)];
            }
            const jsxImport = getJSXRuntimeImport(getJSXImplicitImportBase(options, file), options);
            if (jsxImport) {
                // synthesize `import "base/jsx-runtime"` declaration
                (imports ||= []).push(createSyntheticImport(jsxImport, file));
            }
        }
```

Reply:
I'm not following... To be clear the thing I'm trying to fix is https://twitter.com/cromas/status/1636995012482453505 where changing the setting doesn't make the error go away in the editor and this is the suspected cause; but maybe it isn't?

Reply:
Can you give example error that doesn’t get refreshed ?

Reply:
Hm, I guess I can't reproduce what was described in the issue. e.g.:

![jsx](https://user-images.githubusercontent.com/5341706/236908538-ce785e79-4121-4fc7-97cc-8864d96471fa.gif)

@DanielRosenwasser What were you doing that reproduced the aforementioned tweet?

Reply:
I can't repro in the editor but `--watch` is still busted.

![watch-busted-20230816](https://github.com/microsoft/TypeScript/assets/972891/fa7bee53-6f53-4f00-a098-af4b48570bea)

Reply:
Cool, let me try that instead. Hoping to wrap this one up since it's been sitting for a while 😄

Reply:
er, the action item wasn't on me, I don't think; I think @sheetalkamat was looking for an example of the problem.

Reply:
Ok got it..  can we add this test case as well. This is because the jsx import is not added to file unless its external file and we have some kind of jsxImportPragma so the imports might not get added and hence we need affectsSemanticDiagnostics apart from affectsModuleResolution for the jsx

Reply:
Can you also handle "correct allowJs" value by adding https://github.com/microsoft/TypeScript/commit/236d1b13a8710c61b87c6e23ceec9403b8b61096

Reply:
I'm a bit confused by the above commit; the only flag in that commit that is `allowJsFlag` is `allowJs` itself; should it have been flipped to `checkJsFlag?

Reply:
"allowJs" option is calculated as ` compilerOptions.allowJs === undefined ? !!compilerOptions.checkJs : compilerOptions.allowJs;` 

https://github.com/microsoft/TypeScript/blob/0464e91c8b67579a4ed840e5783575a493c958e0/src/compiler/utilities.ts#L8757

Reply:
Alright, cherry picked, and added a test case for the `jsx` thing in watch mode, which does appear to reproduce the error. For example if I revert my fix and rerun the tests in the PR, I get:

![image](https://github.com/microsoft/TypeScript/assets/5341706/c3d437d9-32e7-4a79-b43a-3b7ea56f65e7)

Adding the bad error back.